lib/step.js

```diff
@@ -167,71 +165,75 @@ class MetaStep extends Step {
     this.actor = obj;
   }
 
-  humanize() {
-    return this.name;
+  /** @return {boolean} */
+  isBDD() {
+    if (this.actor && this.actor.match && this.actor.match(/^(Given|When|Then|And)/)) {
+      return true;
+    }
+    return false;
   }
 
-  setTrace() {
+  isWithin() {
+    if (this.actor && this.actor.match && this.actor.match(/^(Within)/)) {
+      return true;
+    }
+    return false;
   }
 
-  /** @return {void} */
-  run() {
+  toString() {
+    const actorText = !this.isBDD() && !this.isWithin() ? `${this.actor}:` : this.actor;
+    return `${this.prefix}${actorText} ${this.humanize()} ${this.humanizeArgs()}${this.suffix}`;
   }
-}
 
-/** @type {Class<MetaStep>} */
-Step.MetaStep = MetaStep;
+  humanize() {
+    return this.name;
+  }
 
-module.exports = Step;
+  setTrace() {
+  }
 
-function detectMetaStep(stack) {
-  if (store.debugMode) return; // no detection in debug
-  if (!support) loadSupportObjects(); // deprecated
-
-  for (let i = STACK_LINE; i < stack.length; i++) {
-    const line = stack[i].trim();
-    if (isTest(line) || isBDD(line)) break;
-    const fnName = line.match(/^at (\w+)\.(\w+)\s\(/);
-    if (!fnName) continue;
-    if (fnName[1] === 'Generator'
-      || fnName[1] === 'recorder'
-      || fnName[1] === 'Runner'
-    ) { return; } // don't track meta steps inside generators
-
-    if (fnName[1] === 'Object') {
-      // detect PO name from includes
-      for (const name in support) {
-        const file = support[name];
-        if (line.indexOf(file) > -1) {
-          return new MetaStep(`${name}:`, fnName[2]);
-        }
-      }
-    }
-    return new MetaStep(`${fnName[1]}:`, fnName[2]);
+  setContext(context) {
+    this.context = context;
   }
-}
 
-function loadSupportObjects() {
-  const config = require('./config');
-  const path = require('path');
-  support = {
-    ...config.get('include', {}),
-  };
-  if (support) {
-    for (const name in support) {
-      const file = support[name];
-      support[name] = path.join(global.codecept_dir, file);
+  /** @return {*} */
+  run(fn) {
+    this.status = 'queued';
+    this.setArguments(Array.from(arguments).slice(1));
+    let result;
+    // run async before step hooks
+    event.emit(event.metaStep.before, this);
+
+    const registerStep = (step) => {
+      step.metaStep = this;
+    };
+    event.dispatcher.on(event.step.before, registerStep);
+    try {
+      event.emit(event.metaStep.started, this);
+      this.startTime = Date.now();
+      result = fn.apply(this.context, this.args);
+
+      event.emit(event.metaStep.after, this);
+      event.emit(event.metaStep.passed, this);
+    } catch (error) {
+      this.status = 'failed';
+
+      event.emit(event.metaStep.after, this);
+      event.emit(event.metaStep.failed, this);
```

Comment:
I am not sure metaStep.passed/metaStep failed will be actually called, as `fn` is asynchronous, so this try/catch won't show you execution result of this function.
I don't even think `metStep passed/failed` should be implemented at all.